### PR TITLE
feature: add project mirror resources to provider

### PR DESCRIPTION
**Pull request by lukemgriffith**
Implemented functionality for project mirrors, as per issue https://github.com/terraform-providers/terraform-provider-gitlab/issues/150

This was recently added to the go-gitlab library by myself and released in v0.33.0

https://github.com/xanzy/go-gitlab/pull/879

### CODE REVIEW

File: gitlab/resource_gitlab_project_mirror.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+		Create: resourceGitlabMirrorCreate,
+		Read:   resourceGitlabProjectMirrorRead,
+		Update: resourceGitlabMirrorUpdate,
+		Delete: resourceGitlabMirrorDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"project_id": {
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror.go:23]: "project" seems to be more common in this provider than "project_id", although I do prefer "project_id" myself. Is that a conscious choice?

File: gitlab/resource_gitlab_project_mirror.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+		log.Printf("[DEBUG] project mirror found %v", m.ID)
+		if m.ID == mirrorID {
+			mirror = m
+			found = true
+			return nil
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("unable to find mirror %v on project %s", mirrorID, projectID)
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror.go:156]: Instead of returning an error, can you `setId("")`? That seems to be the best practice for handling missing upstream resources.

File: gitlab/resource_gitlab_project_mirror.go (hunk 1 above)
Comment (lukemgriffith) [reply gitlab/resource_gitlab_project_mirror.go:23]: I copied the schema from gitlab project member, so not conscious. Happy to change

File: gitlab/resource_gitlab_project_mirror.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	var mirror *gitlab.ProjectMirror
+	found := false
+
+	for _, m := range mirrors {
+		log.Printf("[DEBUG] project mirror found %v", m.ID)
+		if m.ID == mirrorID {
+			mirror = m
+			found = true
+			return nil
+		} else {
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror.go:152]: Looks like a bug. This for loop will never make it past the first iteration.

**Comment by armsnyder** [comment]
Can you add a test for the importer as well? You can copy an _import test from another resource without much change.

**Comment by lukemgriffith** [comment]
Will do, late GMT so will do tomorrow. Thanks for reviewing and patience much appreciated.

**Comment by lukemgriffith** [comment]
Having issues adding the importer at the moment, and quite make sense of the error either. Still working on this, any input would be great - im not too familiar with the import tests, or the workflow it runs. Trying to pin down where the call to gitlab is being made.

**Comment by armsnyder** [comment]
The importer test is fine. The problem occurs after the import, during the read. Your read function is depending on some attributes that you are not setting in your import function.

Check the docs for `ImportStatePassthrough`

https://www.terraform.io/docs/extend/resources/import.html#importer-state-function

> Where possible, prefer using schema.ImportStatePassthrough as the Importer State function. This function requires the Read function to be able to refresh the entire resource with d.Id() ONLY. Sometimes it is possible to adjust the resource Read function to replace d.Get() use with d.Id() if they exactly match or add a function that parses the resource ID into the necessary attributes.

**Comment by jeremad** [comment]
You can now rebase on master, go-gitlab has been updated

**Comment by lukemgriffith** [comment]
I've got the acceptance tests working, testing manually resource imports fine also.

**Comment by lukemgriffith** [comment]
rebased and caused a lot of mess. Resouce is working as expected. Do you require the commit history to be cleared up any further or is this good?

File: gitlab/resource_gitlab_project_mirror.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+func resourceGitlabProjectMirror() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabMirrorCreate,
+		Read:   resourceGitlabProjectMirrorRead,
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror.go:15]: Inconsistent function naming scheme here

File: gitlab/resource_gitlab_project_mirror.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	return nil
+}
+
+func resourceGitlabProjectMirrorSetToState(d *schema.ResourceData, projectMirror *gitlab.ProjectMirror, projectID *string) {
+
+	mirrorID := strconv.Itoa(projectMirror.ID)
+	d.Set("enabled", projectMirror.Enabled)
+	d.Set("only_protected_branches", projectMirror.OnlyProtectedBranches)
+	d.Set("keep_divergent_refs", projectMirror.KeepDivergentRefs)
+	d.SetId(buildTwoPartID(projectID, &mirrorID))
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror.go:175]: Nitpick, but this line is unnecessary since in all cases the ID will have already been set before calling this function.

File: gitlab/resource_gitlab_project_mirror_test.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
+package gitlab
+
+// TODO convert to handle project mirror.
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror_test.go:3]: I can't tell what this means.

File: gitlab/resource_gitlab_project_mirror_test.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+		CheckDestroy: testAccCheckGitlabProjectMirrorDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccGitlabProjectMirrorConfig(rInt),
+			},
+			{
+				ResourceName:            "gitlab_project_mirror.foo",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"enabled", "mirror_id", "project", "url"},
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror_test.go:85]: Why was this needed? The import should result in a fully synced state, ideally.

File: gitlab/resource_gitlab_project_mirror_test.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+type testAccGitlabProjectMirrorExpectedAttributes struct {
+	URL                   string
+	Enabled               bool
+	OnlyProtectedBranches bool
+	KeepDivergentRefs     bool
+}
+
+func testAccCheckGitlabProjectMirrorAttributes(mirror *gitlab.ProjectMirror, want *testAccGitlabProjectMirrorExpectedAttributes) resource.TestCheckFunc {
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror_test.go:133]: Check your error message strings. They look copy-pasted.

**Comment by armsnyder** [comment]
If you could squash this MR that could be good, because GitHub looks confused. Maybe it's just me but it's showing changes in the diff that aren't yours.

**Comment by lukemgriffith** [comment]
That occured once I performed `git rebase master` and it interlaced the more recent commits. 
I will tidy this up from master and address the more recent comments. 

Re: import test. Testing manually the resource is imported fine using the terraform cli, the import test does not show any fields. I've copied the test from other resources unsure why this is causing problems as the functionality works.

**Comment by lukemgriffith** [comment]
Import test success, I wasn't setting all properties

File: gitlab/resource_gitlab_project_mirror_test.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+			return fmt.Errorf("got only protected branches %t; want %t", mirror.OnlyProtectedBranches, want.OnlyProtectedBranches)
+		}
+		if mirror.KeepDivergentRefs != want.KeepDivergentRefs {
+			return fmt.Errorf("got keep divergent refs %t; want %t", mirror.KeepDivergentRefs, want.KeepDivergentRefs)
+		}
+		return nil
+	}
+}
+
+func testAccCheckGitlabProjectMirrorDestroy(s *terraform.State) error {
```
Comment (armsnyder) [inline gitlab/resource_gitlab_project_mirror_test.go:149]: Just an fyi, a "destroy" test should be testing the behavior of your own resource destroy function, i.e. checking that the mirror gets disabled. There's an open issue about this generally #205 so we can come back and do this later.

**Comment by armsnyder** [comment]
Can you add a new documentation page for this new resource?

**Comment by lukemgriffith** [comment]
added docs page.

File: gitlab/resource_gitlab_project_mirror_test.go (hunk 9 above)
Comment (lukemgriffith) [reply gitlab/resource_gitlab_project_mirror_test.go:149]: sure, I will not make any changes in this PR for this.

**Comment by armsnyder** [comment]
Looks good, but can you add the new doc page to the index (like so https://github.com/gitlabhq/terraform-provider-gitlab/pull/391)

**Comment by lukemgriffith** [comment]
ack i actually searched for an index, didn't realise it was the erb file. should be done.

**Comment by lukemgriffith** [comment]
@armsnyder is there anything I can do to speed this up getting merged? I have some work in the next week that i'd be keen to use this resource for... unsure on the release cadence or the process?

**Comment by armsnyder** [comment]
I don't have merging powers, but @nicholasklick may be able to give an update.

**Comment by lukemgriffith** [comment]
Hi @nicholasklick  any timeframe on this getting merged?
